### DISCUSSION TITLE
Jog or speed based control

### DISCUSSION THREAD

**Discussion by rmoscoloni** [gin66/FastAccelStepper#201] [General]
I want to thank gin66 and everyone who contributed to this great library. I want to know how I can use it to make a movement whose speed is related to a signal, for example an analog input (or a digital pulse width) , from 0 to some speed, respecting the acceleration and deceleration ramps. Calling the runFoward() command in each loop cycle works on the acceleration ramp but skips the deceleration ramp.

```
void loop() {
  // this code reads a signal period in an ISR results in speed in hz
  if (micros() - input_micros_last > 23000) {
    input_result = input_read_actual ? 0.585*20512 : 0; //HARCODED 20512 input period
  } else
    input_result = 0.585*input_period;
  
  //if(!stepper->isRunning()) 
  //  stepper->runForward(); 
  
  stepper->runForward();

  if (duty_result == 0 )
    stepper->stopMove();
  else
  stepper->setSpeedInHz(duty_result);
  }
```

**Comment by gin66** [comment]
Perhaps use `runForward()` from standstill or if `duty_result != 0` like this ?

```
void loop() {
  // this code reads a signal period in an ISR results in speed in hz
  if (micros() - input_micros_last > 23000) {
    input_result = input_read_actual ? 0.585*20512 : 0; //HARCODED 20512 input period
  } else
    input_result = 0.585*input_period;
  
  if(!stepper->isRunning() || (duty_result != 0)) 
    stepper->runForward(); 
  
  if (duty_result == 0 )
    stepper->stopMove();
  else {
      stepper->setSpeedInHz(duty_result);
      // the new value is applied on runForward() above
  }
 }
```
or shorter, which makes it understandable, that the `isRunning()`-test is not needed:
```
void loop() {
  // this code reads a signal period in an ISR results in speed in hz
  if (micros() - input_micros_last > 23000) {
    input_result = input_read_actual ? 0.585*20512 : 0; //HARCODED 20512 input period
  } else
    input_result = 0.585*input_period;
  
  if (duty_result == 0 )
    stepper->stopMove();
  else {
      stepper->setSpeedInHz(duty_result);
      // the new value is applied by runForward(). If the stepper is not running, it will be started
     stepper->runForward();  
  }
}
```

**Comment by rmoscoloni** [comment]
Thanks for the answer, I didn't want to bother you on a Sunday, your last code works similar to mine, the acceleration ramp works wonderfully, the deceleration ramp doesn't work (it stops abruptly at the new target speed) if the value is different from 0.
 If it is 0, it decelerates normally until it stops.

**Reply by gin66** [reply]
Which library version do you use ? A fix for similar issue is included in 0.30.4.

**Reply by gin66** [reply]
If you use Arduino IDE, perhaps this  [Arduino bug](https://github.com/arduino/libraries-repository-engine/issues/194) affects you, too.

**Comment by rmoscoloni** [comment]
I was on vsc/platformio 0.30.0 to avoid the compiller complain(on 0.30.5) about deprecated rmt_memory_rw_rst, my board is an nodemcu esp32s.
i tested 0.30.5 again, also complain but the deaccel ramp works good!
Thanks for take time to answer, the code is a piece of art!
